kernel/rtlil.cc

```diff
@@ -311,20 +325,30 @@ RTLIL::Const RTLIL::Const::from_string(const std::string &str)
 	return c;
 }
 
-std::string RTLIL::Const::decode_string() const
-{
-	std::string string;
-	string.reserve(GetSize(bits)/8);
-	for (int i = 0; i < GetSize(bits); i += 8) {
-		char ch = 0;
-		for (int j = 0; j < 8 && i + j < int (bits.size()); j++)
-			if (bits[i + j] == RTLIL::State::S1)
-				ch |= 1 << j;
-		if (ch != 0)
-			string.append({ch});
-	}
-	std::reverse(string.begin(), string.end());
-	return string;
+std::string RTLIL::Const::decode_string() const {
+        const int n = GetSize(bits);
+        const int n_over_8 = n / 8;
+        std::string s;
+        s.reserve(n_over_8);
+
+        auto fill_byte = [&](int start, int end) {
+          char ch = 0;
+          for (int j = start; j < end; j++) {
+            if (bits[j] == RTLIL::State::S1) {
+              ch |= 1 << (j - start);
+            }
+          }
+          if (ch != 0) s.append({ch});
+        };
+
+        int i = n_over_8 * 8;
+        if (i < n) {
+          fill_byte(i, n);
```

Comment:
I must say I really dislike the closure here. What about something like:
```
    int cap = GetSize(bits); // where the current byte ends
    for (int i = GetSize(bits) / 8 * 8; i >= 0; i -= 8) {
        for (int j = i; j < cap; j++) {
            ...
        }
        cap = j;
    }

Reply:
Having a helper function for filling each byte is more readable to me, but I suppose it's a matter of what style you are used to read. This is a pretty common idiom in Google code. But we can certainly change it.

Reply:
Please do change it. While at it, please fix the patch to use tabs like in the rest of the file.

Reply:
Also to consider is splitting the patches: we are ready to merge the `decode_string()` change but we are not so sure about the lookup table in the constructor.

Reply:
@QuantamHD do you want to update / split the PR or should I just file a new one?